File: python/codeql/path_injection/benign.py
from pathlib import Path

from flask import Flask, request


app = Flask(__name__)
FILES = {
    "welcome": Path("/srv/app/static/welcome.txt"),
    "status": Path("/srv/app/static/status.txt"),
}


@app.get("/read-safe")
def read_safe():
    path = FILES.get(request.args.get("file", "welcome"), FILES["welcome"])
    return path.read_text(encoding="utf-8")

File: javascript/codeql/path_injection/benign.js
const express = require("express");
const fs = require("fs");

const app = express();
const files = new Map([["welcome", "/srv/app/static/welcome.txt"], ["status", "/srv/app/static/status.txt"]]);

app.get("/read-safe", (req, res) => {
  const file = files.get(String(req.query.file || "welcome")) || files.get("welcome");
  res.send(fs.readFileSync(file, "utf8"));
});
